### PR TITLE
pollfd fix, round 2

### PR DESCRIPTION
**Pull request by vtjnash**
this is a rewrite of the poll_fd, poll_file, and watch_file APIs to more directly expose the libuv API and eliminate the callback legacy interface. since `FDWatcher` objects are now refcounted, this also solves #4401.

this replaces https://github.com/JuliaLang/julia/pull/7275

ping: @Keno

### CODE REVIEW

**Comment by vtjnash** [comment]
cross-ref: this will resolve some https://github.com/JuliaLang/julia/issues/10293

**Comment by vtjnash** [comment]
this closes https://github.com/JuliaLang/julia/issues/7840 (when merged)

File: base/stream.jl [hunk 1]
```diff
@@ -595,6 +597,7 @@ function _uv_hook_close(t::Timer)
     unpreserve_handle(t)
     disassociate_julia_struct(t)
     t.handle = C_NULL
+    notify(t.cond)
```
Comment (vtjnash) [inline base/stream.jl]: this fix is the cause of the warning in the `repl.jl` test: the Timer callback on line 100 of test/repl.jl is supposed to accept an argument `t`

File: base/poll.jl [hunk 2]
```diff
@@ -56,293 +26,381 @@ const FD_TIMEDOUT = 4
...
     handle::Ptr{Void}
-    file::AbstractString
-    open::Bool
+    file::ByteString
+    interval::UInt32
     notify::Condition
-    cb::Callback
-    function PollingFileWatcher(cb, file)
+    active::Bool
+    function PollingFileWatcher(file::AbstractString, interval::Float64=5.007) # same default as nodejs
```
Comment (JeffBezanson) [inline base/poll.jl]: What a weird constant!

File: base/poll.jl [hunk 3]
```diff
@@ -56,293 +26,381 @@ const FD_TIMEDOUT = 4
...
+        if interval < 2
+            warn("For maximum portability, use multi-second intervals with PollingFileWatcher")
```
Comment (JeffBezanson) [inline base/poll.jl]: Could we either make this an error, or give an accurate error based on what the current platform actually supports?

File: base/poll.jl (hunk 2 above)
Comment (vtjnash) [inline base/poll.jl]: i don't know where it comes from either (https://nodejs.org/api/fs.html#fs_fs_watchfile_filename_options_listener)

File: base/poll.jl (hunk 3 above)
Comment (vtjnash) [inline base/poll.jl]: i just copied the warning from the libuv docs (http://docs.libuv.org/en/latest/fs_poll.html#c.uv_fs_poll_start)

File: doc/stdlib/io-network.rst [hunk 4]
```diff
@@ -723,22 +723,33 @@ Network I/O
...
-.. function:: poll_file(s, interval_seconds::Real, seconds::Real)
+   Internally, this uses a ``FDWatcher``.
 
-   Monitor a file for changes by polling every `interval_seconds` seconds for `seconds` seconds. A return value of true indicates
-   the file changed, a return value of false indicates a timeout.
+.. function:: poll_file(path, interval_seconds::Real, timeout_s::Real) -> (previous::StatStruct, current::StatStruct)
+
+   Monitor a file for changes by polling every ``interval_seconds`` seconds until a change occurs or ``timeout_s`` seconds have elapsed.
+
+   Returns a pair of ``StatStruct`` objects when a change is detected. (using ``watch_file`` is preferred, if supported)
```
Comment (JeffBezanson) [inline doc/stdlib/io-network.rst]: This has a very different return value from `watch_file`, so you can't just switch from one to the other. At the same time, if `watch_file` is better, why does `poll_file` exist? Can we pick for you? If not, the docs should describe the tradeoffs so you can pick.

File: doc/stdlib/io-network.rst [hunk 5]
```diff
@@ -723,22 +723,33 @@ Network I/O
...
+   The keyword arguments determine which of read and/or write status should be monitored; at least one of them must be set to true.
 
-   Poll a file descriptor fd for changes in the read or write availability and with a timeout given by the second argument.
-   If the timeout is not needed, use ``wait(fd)`` instead. The keyword arguments determine which of read and/or write status
-   should be monitored and at least one of them needs to be set to true.
    The returned value is an object with boolean fields ``readable``, ``writable``, and
    ``timedout``, giving the result of the polling.
 
-.. function:: poll_file(s, interval_seconds::Real, seconds::Real)
+   Internally, this uses a ``FDWatcher``.
```
Comment (JeffBezanson) [inline doc/stdlib/io-network.rst]: Do you need to know about this type? Is there anything useful you can do with it?

File: doc/stdlib/io-network.rst (hunk 4 above)
Comment (vtjnash) [inline doc/stdlib/io-network.rst]: this was also copied from nodejs: https://nodejs.org/api/fs.html#fs_caveats (although i guess there, they are nice enough to tell you to do a comparison of `curr.mtime and prev.mtime` to make this compatible)

File: doc/stdlib/io-network.rst (hunk 5 above)
Comment (vtjnash) [inline doc/stdlib/io-network.rst]: you could create these types manually and iterate a wait loop somewhat more efficiently (at the risk of losing events, since they are edge-based)

**Comment by JeffBezanson** [comment]
Awesome, this is a relief!

I believe the only piece of #10293 left is SingleAsyncWork?

File: base/poll.jl (hunk 3 above)
Comment (JeffBezanson) [inline base/poll.jl]: Humph, typical wishy-washy nonsense. Also, it says to use multi-second intervals, but that sub-second intervals are the problematic ones. Ambiguous for values between 1 and 2. I think we should just give an error for values less than 1.

File: doc/stdlib/io-network.rst (hunk 5 above)
Comment (JeffBezanson) [inline doc/stdlib/io-network.rst]: Well, we need to make a call here. If you feel using these types directly is not a great idea, we should just leave them as an implementation detail.

Comment (vtjnash) [inline doc/stdlib/io-network.rst]: it's probably best to just consider them an implementation detail for now then. it's easier to revist that later than the reverse.

File: base/poll.jl (hunk 3 above)
Comment (vtjnash) [inline base/poll.jl]: the libuv implementation is to call `stat` in a loop off a uv_timer_t and invoke the callback if any of the fields are not identical. other than being extremely inefficient for small times (e.g, from what I could tell, osx was only giving me 1sec resolution for mtime), I actually don't see an issue.

Comment (JeffBezanson) [inline base/poll.jl]: I just prefer not to annoy people with warnings. Realistically, somebody is not going to happily run a program that scrolls that warning repeatedly, so it might as well be an error. Especially since the warning depends directly on a passed argument, we need to decide whether the argument value is allowed or not.

File: base/libc.jl [hunk 6]
```diff
@@ -23,6 +23,15 @@ dup(src::RawFD,target::RawFD) = systemerror("dup",-1==
     ccall((@windows? :_dup2 : :dup2),Int32,
     (Int32,Int32),src.fd,target.fd))
 
+#Wrapper for an OS file descriptor (for Windows)
+@windows_only immutable WindowsRawSocket
+    handle::Ptr{Void}   # On Windows file descriptors are HANDLE's and 64-bit on 64-bit Windows...
+end
+
+@unix_only _get_osfhandle(fd::RawFD) = fd
+@windows_only _get_osfhandle(fd::RawFD) = WindowsRawSocket(ccall(:_get_osfhandle,Ptr{Void},(Cint,),fd.fd))
```
Comment (tkelman) [inline base/libc.jl]: this'll need to be qualified with `Libc.` now anywhere it's used, like `base/stream.jl` and `base/mmap.jl`

File: test/pollfd.jl [hunk 7]
```diff
@@ -1,52 +1,90 @@
 # This file is a part of Julia. License is MIT: http://julialang.org/license
 
-@unix_only begin
```
Comment (tkelman) [inline test/pollfd.jl]: why did you remove this? msvcrt has `_pipe` but it has extra arguments: https://msdn.microsoft.com/en-us/library/aa298531(v=vs.60).aspx

Comment (vtjnash) [inline test/pollfd.jl]: the eventual goal is to get this test running on windows too

Comment (tkelman) [inline test/pollfd.jl]: not a bad goal, but will need to account for different spellings/api of posix functions - unless libuv has wrappers for all of these, which I think is one of the things we've been meaning to do upstream?

**Comment by vtjnash** [comment]
bump. this passes CI everywhere now...

File: doc/stdlib/io-network.rst [hunk 8]
```diff
@@ -723,22 +723,35 @@ Network I/O
...
    The returned value is an object with boolean fields ``readable``, ``writable``, and
    ``timedout``, giving the result of the polling.
 
-.. function:: poll_file(s, interval_seconds::Real, seconds::Real)
+.. function:: poll_file(path, interval_seconds::Real, timeout_s::Real) -> (previous::StatStruct, current::StatStruct)
 
-   Monitor a file for changes by polling every `interval_seconds` seconds for `seconds` seconds. A return value of true indicates
-   the file changed, a return value of false indicates a timeout.
+   Monitor a file for changes by polling every ``interval_seconds`` seconds until a change occurs or ``timeout_s`` seconds have elapsed.
+   The `timeout_s` should be a long interval; the default is 5.007 seconds.
```
Comment (JeffBezanson) [inline doc/stdlib/io-network.rst]: Should say interval instead of timeout_s.

File: src/support/wsasocketpair.c [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+// since this is not a Win32 API call
+// (despite MSDN documentation to the contrary,
+// it will not work with ReadFile/WriteFile)
+
+#define WIN32_LEAN_AND_MEAN
+#include <winsock2.h>
+#include <io.h>
+#define AF_UNIX 1
+
+__declspec(dllexport) int wsasocketpair(int domain, int type, int protocol, SOCKET socket_vector[2]) {
```
Comment (JeffBezanson) [inline src/support/wsasocketpair.c]: Is this file only used for testing? Will it be useful for anything else?

**Comment by ViralBShah** [comment]
Marking this as 0.4 - assuming this is is done and long standing, and we should get it in.

File: src/support/wsasocketpair.c (hunk 9 above)
Comment (vtjnash) [inline src/support/wsasocketpair.c]: from our libuv fork, we have the equivalent of `pipe` that returns NamedPipe objects. this provides the ability to have the equivalent of `socketpair` instead. I suspect both would be useful to get pushed upstream into libuv.

**Comment by vtjnash** [comment]
bump. i'll merge this later today.

**Comment by amitmurthy** [comment]
Awesome. :+1:

**Comment by tkelman** [comment]
travis failure here https://gist.github.com/tkelman/96b21860b35454115373, are we back to timings that are too picky for busy low-memory VM's?

**Comment by amitmurthy** [comment]
OTOH, in this case it does seem like it caught an actual bug. The poll call should not have returned before `intvl` seconds.

**Comment by tkelman** [comment]
Another failure: https://travis-ci.org/JuliaLang/julia/jobs/73427729

```
exception on 5: ERROR: LoadError: test failed: 0.1 <= 0.099903232
 in expression: intvl <= t_elapsed
 in yieldto at ./task.jl:24
 in wait at ./task.jl:313
 in wait at ./task.jl:228
 in wait at ./task.jl:61
 in sync_end at ./task.jl:340
 in anonymous at no file:348
 in include_string at loading.jl:158
 in include_from_node1 at ./loading.jl:200
 in runtests at /tmp/julia/share/julia/test/testdefs.jl:173
 in anonymous at multi.jl:909
 in run_work_thunk at multi.jl:648
 in anonymous at task.jl:909
while loading /tmp/julia/share/julia/test/pollfd.jl, in expression starting on line 65
	From worker 3:	     * bitarray              in  90.72 seconds
	From worker 4:	     * mpfr                  in   4.91 seconds
	From worker 2:	     * subarray              in 353.43 seconds
ERROR: LoadError: LoadError: test failed: 0.1 <= 0.099903232
 in expression: intvl <= t_elapsed
 in anonymous at task.jl:1442
while loading /tmp/julia/share/julia/test/pollfd.jl, in expression starting on line 65
while loading /tmp/julia/share/julia/test/runtests.jl, in expression starting on line 5
	From worker 5:	     * pollfd          
```

**Comment by tkelman** [comment]
another pollfd failure on travis https://s3.amazonaws.com/archive.travis-ci.org/jobs/75113270/log.txt

**Comment by timholy** [comment]
Another pollfd failure on AppVeyor:
image
in #12560.

**Comment by vtjnash** [comment]
i'm not really sure what that test is doing in the pollfd.jl file, since it is testing the functionality of one-shot Timers, not any of this code

@JeffBezanson @Keno is 0.2 sec sometimes not enough for a 0.01 sec timer to trigger? (added in 7f4c2de4ee7ac6403386b8d67c78826a4db36859)

**Comment by tkelman** [comment]
same `Any[] == [1]` pollfd failure on a buildbot http://buildbot.e.ip.saba.us:8010/builders/build_ubuntu14.04-x86/builds/2163/steps/shell_2/logs/stdio

**Comment by vtjnash** [comment]
latest travis failure on #12463 (https://travis-ci.org/JuliaLang/julia/builds/76174785) was also due to Jeff's new test for Timers there
